src/on_shutdown.c

```diff
@@ -0,0 +1,182 @@
+#include "on_shutdown.h"
+
+#include <stdlib.h>
+#include <small/rlist.h>
+#include <trivia/util.h>
+#include <fiber.h>
+#include <libev/ev.h>
+
+struct on_shutdown_trigger {
+	/*
+	 * rlist entry
+	 */
+	struct rlist entry;
+	/*
+	 * Shutdown handler arg
+	 */
+	void *arg;
+	/*
+	 * Shutdown handler
+	 */
+	void (*handler)(void *);
+	/*
+	 * Async event set when module main thread finished
+	 */
+	struct ev_async on_shutdown_async;
+	/*
+	 * Module shutdown fiber
+	 */
+	struct fiber *on_shutdown_fiber;
+	/*
+	 * Main loop
+	 */
+	struct ev_loop *loop;
+};
+
+/*
+ * Shutdown trigger list, not need mutex, because access is
+ * avaliable only from tx thread
+ */
+static RLIST_HEAD(on_shutdown_trigger_list);
+
+static void
+on_shutdown_fiber_wakeup(MAYBE_UNUSED struct ev_loop *loop, ev_async *ev, MAYBE_UNUSED int revents)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger = ev->data;
+	fiber_wakeup(on_shutdown_trigger->on_shutdown_fiber);
+}
+
+static int
+create_new_on_shutdown_trigger(void *arg, void (*handler)(void *), void **opaque)
+{
+	struct on_shutdown_trigger *trigger = (struct on_shutdown_trigger *)
+		malloc(sizeof(struct on_shutdown_trigger));
+	if (trigger == NULL)
+		return -1;
+	trigger->arg = arg;
+	trigger->handler = handler;
+	ev_async_init(&trigger->on_shutdown_async, on_shutdown_fiber_wakeup);
+	trigger->on_shutdown_async.data = trigger;
+	*opaque = trigger;
+	trigger->loop = loop();
+	rlist_add_entry(&on_shutdown_trigger_list, trigger, entry);
+	return 0;
+}
+
+USED int
+on_shutdown_register(void *arg, void (*new_hadler)(void *), void (*old_handler)(void *),
+	void **opaque)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger, *tmp;
+	if (old_handler == NULL) {
+		if (new_hadler != NULL) {
+			return create_new_on_shutdown_trigger(arg, new_hadler, opaque);
+		} else {
+			/*
+			 * Invalid function params, old_handler or new_hadler
+			 * must be set
+			 */
+			return -1;
```

Comment:
Dividing this function into two ones (create() and delete()) would allow you to remove this extra check and replace it with an assertion.